slapo/sharding/utils.py

```diff
@@ -0,0 +1,152 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+from __future__ import annotations
+
+import torch
+import torch.distributed as dist
+
+from ..logger import get_logger
+
+logger = get_logger()
+
+
+def all_gather_along_dim(inp, dim, rank, world_size, group):
+    """all-gather along the given dimension. Will use all_gather_into_tensor
+    if available as it is more efficient; otherwise fallback to all_gather.
+
+    Paramters
+    ---------
+    inp: torch.Tensor
+        The input tensor to all-gather.
+    dim: int
+        The dimension to all-gather along.
+    rank: int
+        The rank of the current process.
+    world_size: int
+        The number of processes in the group.
+    group: torch.distributed.ProcessGroup
+        The process group to all-gather.
+
+    Returns
+    -------
+    torch.Tensor
+        The gathered tensor.
+    """
+    if hasattr(dist, "all_gather_into_tensor"):
+        temp = inp.transpose(0, dim).contiguous() if dim != 0 else inp
+        gather_shape = list(temp.shape)
+        gather_shape[0] = world_size * gather_shape[0]
+        ret = torch.empty(gather_shape, dtype=temp.dtype).cuda(rank)
+        dist.all_gather_into_tensor(ret, temp, group=group)
+        ret = ret.transpose(0, dim).contiguous() if dim != 0 else ret
+    else:
+        # Fallback to all_gather. This may lead to suboptimal performance.
+        parts = [
+            torch.empty(inp.shape, dtype=inp.dtype).cuda(rank)
+            for _ in range(world_size)
+        ]
+        dist.all_gather(parts, inp, group=group)
+        ret = torch.cat(parts, dim=dim)
+    return ret
+
+
+class _AllGatherForwardOutput(torch.autograd.Function):
+    """The cusom all gather op used for forward hook."""
+
+    # pylint: disable=abstract-method, arguments-differ
+    @staticmethod
+    def forward(ctx, inp, dim, group):
+        ctx.dim = dim
+        ctx.group = group
+        world_size = dist.get_world_size(group)
+        rank = dist.get_rank(group)
+        return all_gather_along_dim(inp, dim, rank, world_size, group)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        dim = ctx.dim
+        group = ctx.group
+        world_size = dist.get_world_size(group)
+        rank = dist.get_rank(group)
+        sharded_size = grad_output.shape[dim] // world_size
+        ret = grad_output.split(sharded_size, dim=dim)[rank]
```

Comment:
we would need contiguous when dim is the primary dimension.

Reply:
Fixed.